### ISSUE TITLE
domvm.onevent

### ISSUE THREAD

**Issue by leeoniya** [leeoniya/domvm#121]
i'm thinking of adding a global `domvm.onevent` hook to mini+ builds to augment the "view reactivity" story. basically it would provide a way to hook into all event handlers in the app and allow you to add a root auto-redraw or whatever.

due to it being global, this is something that can be used by apps at the top level, but should not be used by reusable/composable components.

it would default to a no-op, but can be overwritten to something like...

```js
domvm.onevent = function(e, result, vm, node) {
    rootVm.redraw();
};
```

`result` is the return value of the bound event handler, so could be a Promise if the event resulted in eg. an async xhr/fetch request, so any `redraw()` can be chained to it, as needed. not sold on the argument order yet, but seems sane enough.

**Comment by lawrence-dol** [comment]
Seems like a good idea. I think the arg order might be better as event, node, vm, result. Which is to say the arguments should be consistent with the order given to an event handler.

**Comment by leeoniya** [comment]
i generally like to order args by their usefulness within the context. in this instance i think e, result and vm are most useful, with node being quite useless. but consistency is good too :/

**Comment by lawrence-dol** [comment]
> consistency is good

For me, consistency is king (for this kind of decision).

**Comment by leeoniya** [comment]
well, there's still a bit of a gotcha here.

parameterized event handlers `onclick: [myFn, 1, 2 3]` get called as `myFn(1, 2, 3, e, node, vm)` [1]. so if you wanted to handle any event globally in a fashion that's uniform with the regular handler semantics, you'd have to deal with detecting which sig the global callback is invoked with by locating `e` in the arguments.

[1] https://github.com/leeoniya/domvm/blob/2.x-dev/src/view/patchEvent.js#L12

**Comment by lawrence-dol** [comment]
And I guess it's too late to change that to `[e, node, getVm(node)].concat(args)`. Shame. I would have found the latter far more intuitive than what's there now.

**Comment by leeoniya** [comment]
the thought process was that in parameterized handlers, you almost certainly care more about the explicit args than the automatic ones added by domvm. that way your handler sigs need not always worry about auxiliary data.

it's too late to change this in v2, but TBH i don't think it's that big an issue besides being non-uniform, since  your handlers are originally defined using a non-uniform syntax. the syntax of the definition matches that of the callback sig - which is a good balance of predictability and practicality for me.

the lack of global uniformity does impact this specific application though. again, arg-testing isnt drastically expensive, unless a user's doing stupid things like non-throttled scroll/mousemove/resize events.

**Comment by leeoniya** [comment]
okay, a global `onevent` should definitely have a consistent signature; arg-testing would be a giant clusterfuck. i still want to pass through parameterized args if they're provided, so `[e, node, getVm(node)].concat(args)` looks like a winner. i'll consider changing the parameterized event handler sig to match this in a future version,

**Comment by leeoniya** [comment]
implemented in 0d9513fcf441fadbc059d2ba45db0309453e0de1 as `domvm.config({onevent: (e, node, vm, ...args) {...}})`. demo: https://rawgit.com/leeoniya/domvm/2.x-dev/demos/global-onevent.html
